http-client/src/main/java/io/micronaut/http/client/netty/Http1ResponseHandler.java

```diff
@@ -0,0 +1,355 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.client.netty;
+
+import io.micronaut.http.body.CloseableByteBody;
+import io.micronaut.http.client.exceptions.ResponseClosedException;
+import io.micronaut.http.netty.body.AvailableNettyByteBody;
+import io.micronaut.http.netty.body.BodySizeLimits;
+import io.micronaut.http.netty.body.BufferConsumer;
+import io.micronaut.http.netty.body.StreamingNettyByteBody;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.ReferenceCountUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Inbound message handler for the HTTP/1.1 client. Also used for HTTP/2 and /3 through message
+ * translators.
+ *
+ * @author Jonas Konrad
+ * @since 4.7.0
+ */
+final class Http1ResponseHandler extends SimpleChannelInboundHandlerInstrumented<HttpObject> implements BufferConsumer.Upstream {
+    private static final Logger LOG = LoggerFactory.getLogger(Http1ResponseHandler.class);
+
+    private final ResponseListener listener;
+    private State state = State.BEFORE_RESPONSE;
+    private HttpResponse response;
+    private List<ByteBuf> buffered;
+    private StreamingNettyByteBody.SharedBuffer streaming;
+    private long demand;
+    private ChannelHandlerContext streamingContext;
+
+    public Http1ResponseHandler(ResponseListener listener) {
+        super(false);
+        this.listener = listener;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.read();
+    }
+
+    @Override
+    protected void channelReadInstrumented(ChannelHandlerContext ctx, HttpObject msg) throws Exception {
+        if (msg.decoderResult().isFailure()) {
+            ReferenceCountUtil.release(msg);
+            exceptionCaught(ctx, msg.decoderResult().cause());
+            return;
+        }
+
+        HttpContent content;
+        switch (state) {
+            case BEFORE_RESPONSE:
+                HttpResponse response = (HttpResponse) msg;
+                if (response.status().code() == HttpResponseStatus.CONTINUE.code()) {
+                    listener.continueReceived(ctx);
+                    if (response instanceof LastHttpContent lhc) {
+                        lhc.release();
+                    } else {
+                        assert !(response instanceof HttpContent);
+                        state = State.DISCARDING_CONTINUE_CONTENT;
+                    }
+                    break;
```

Comment:
This code looks complex because of the different state fields and switch breaks, which make it hard to follow the flow. I would represent the state as an object that can react to read, exception, discard, etc. Each state would contain only the fields it needs.

pseudocode:

if (state == null) {
    assume response
    state = new BUFFERED_CONTENT()
}

state = state.read(...)


void channelReadComplete() {
    state.readComplete()
}

Reply:
that sounds like the PipeliningServerHandler approach. will try

Reply:
pushed, it looks quite different now.

Reply:
Can you make the classes static and set the state only in the handler?

Reply:
Have the methods in `ReaderState` always return `ReaderState`?

Reply:
No it's not possible. There are some necessary ordering conditions. For example, BufferedContent must update the state before listener.complete is called, because listener.complete may trigger further operations that need to be sent to the new state.

Reply:
I thought that the handler could not be accessed in parallel, so the method has to finish before a second invocation of any method on a handler or a listener

Reply:
netty allows no multi-threading for a handler but it does allow reentrancy. e.g. if you do ctx.read inside a handler method, channelRead of the same handler may be called inside the ctx.read call. It's less strict than reactor in this regard.

Reply:
I see